# Bio-JEPA AC Explainer


In [1]:
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch.utils.data import Dataset, DataLoader

import copy
import math
import numpy as np
import glob
import os
import json
from pathlib import Path
import matplotlib.pyplot as plt
import random
from dataclasses import dataclass

In [2]:
torch.manual_seed(1337)
random.seed(1337)

In [3]:
data_dir = Path('/Users/djemec/data/jepa')
shard_dir = data_dir / 'tokenized'
train_dir = shard_dir / 'train'
val_dir =  shard_dir / 'val'
metadata_path = data_dir / 'perturbation_map.json'
checkpoint_dir = data_dir / 'checkpoint'

## Text Prep/Tokenization

We previously loaded and saved our data into a tokenized files.  For a full explainer, look at the data_prep_explainer.ipynb notebook.   In this step we'll pull the data from the files to use for training.  First we'll start by getting all the files in the tokenized directory. 

In [4]:
data_files = sorted(train_dir.glob('*.npz'))
print(f'Found {len(data_files)} shards.')

Found 29 shards.


Now that we have the files, we need just a single file. Because of how we tokanized our files, we do not have to load them in any particular order so we'll select a random file from the list.  In training, we'd want to remove this from the file list to ensure we go through the full epoch before restarting.   

In [5]:
random_file_idx = np.random.randint(len(data_files))
file_path = data_files[random_file_idx]
random_file_idx, file_path

(28, PosixPath('/Users/djemec/data/jepa/tokenized/train/shard_0029.npz'))

# Modeling

## Data Loading


To start, we need to get enough data to run the forward and backward passes.  Since our total dataset is likely too big to be held in memory all at once in real practice, we will read just enough file information into memory so that we can run the passes, leaving memory and compute to be used on the passes instead of static data holding. 

Recall that we saved the data in the following structure:

```
{
    'control':[array of network expression information],
    'case':[array of network expression information],
    'action_ids':[array of ids of the perturbation]
}
```

Because of this, taking the same index out of each array gives us our action_id, pertrubed data, and control data. For training we'll do more than 1 cell at a time.  We'll use a variable `BATCH_SIZE` to control how much data to train on at a time.  In this case we'll start with 2. During training we'll want a increase our batch size.  


In [6]:
BATCH_SIZE = 2
with np.load(file_path) as data:
    # Keys: 'control', 'treated', 'action_ids'
    n_rows = data['action_ids'].shape[0]
    row_idx = np.random.randint(n_rows, size=BATCH_SIZE)
    
    control_raw = data['control'][row_idx]
    case_raw = data['case'][row_idx]
    act_id = data['action_ids'][row_idx]

x_control = torch.tensor(control_raw)
x_case = torch.tensor(case_raw)
action_id = torch.tensor(act_id, dtype=torch.long)

row_idx, action_id, x_control, x_case

(array([1517, 1699]),
 tensor([1383,  429]),
 tensor([[0.7684, 0.6221, 0.5242,  ..., 0.3232, 0.2548, 0.4801],
         [0.5225, 0.5175, 0.4678,  ..., 0.3220, 0.2001, 0.4899]]),
 tensor([[0.6076, 0.4994, 0.4782,  ..., 0.3042, 0.2252, 0.5015],
         [0.7085, 0.4856, 0.5094,  ..., 0.3580, 0.1925, 0.4837]]))

## Forward pass

![network_impact_math](../resources/biojepa_v1.png)

Our current architecture has 3 models:  

1. **The Student** - starts from the control cell and learns how to create a latent embedding of the cell state. 
2. **The Predictor** - learns how to adjust the student's embedded cell state based on different perturbations to hallucinate the future diseased state in the latent space shared with the teacher. 
3. **The Teacher** - starts from the actual perturbed cell and creates a latent space embedding for the pertured state to grade the predictors output against.


We'll start by building the Pre-Norm Transformer Encoder block with Rotary Positional Embeddings (RoPE) that are the basis for both the Teacher and Student models. 

### Teacher/Student Setup
**RoPE**

In [7]:
class RotaryEmbedding(nn.Module):
    def __init__(self, dim, max_seq_len=2048):
        super().__init__()
        # Ensure dim is the head dimension, not the full embedding dimension
        inv_freq = 1.0 / (10000 ** (torch.arange(0, dim, 2).float() / dim))
        t = torch.arange(max_seq_len).type_as(inv_freq)
        freqs = torch.outer(t, inv_freq)
        emb = torch.cat((freqs, freqs), dim=-1)
        self.register_buffer('emb', emb)

    def forward(self, x):
        # x: [Batch, Seq, Dim]
        # We only care about the sequence length here
        n = x.shape[1]
        # Returns [Seq, Head_Dim]
        return self.emb[:n, :].cos(), self.emb[:n, :].sin()

def apply_rotary_pos_emb(q, k, cos, sin):
    # q, k: [Batch, Heads, Seq, Head_Dim]
    # cos, sin: [Seq, Head_Dim] -> reshape to [1, 1, Seq, Head_Dim]
    
    cos = cos.unsqueeze(0).unsqueeze(0)
    sin = sin.unsqueeze(0).unsqueeze(0)

    # Standard RoPE rotation logic
    # split last dim into half
    q_d = q.shape[-1] // 2
    k_d = k.shape[-1] // 2
    
    q1, q2 = q[..., :q_d], q[..., q_d:]
    k1, k2 = k[..., :k_d], k[..., k_d:]
    
    q_rotated = torch.cat((-q2, q1), dim=-1)
    k_rotated = torch.cat((-k2, k1), dim=-1)
    
    q_out = (q * cos) + (q_rotated * sin)
    k_out = (k * cos) + (k_rotated * sin)
    
    return q_out, k_out

**Multi-Headed Attention**

In [8]:
class BioMultiHeadAttention(nn.Module):
    # mirrors nn.MultiheadAttention(dim, heads, batch_first=True) 
    def __init__(self, config):
        super().__init__()
        assert config.embed_dim % config.heads == 0
        
        self.head_dim = config.embed_dim // config.heads
        self.heads = config.heads
        self.embed_dim = config.embed_dim
        
        # Projections
        self.q_proj = nn.Linear(config.embed_dim, config.embed_dim)
        self.k_proj = nn.Linear(config.embed_dim, config.embed_dim)
        self.v_proj = nn.Linear(config.embed_dim, config.embed_dim)
        
        self.c_proj = nn.Linear(config.embed_dim, config.embed_dim)

    def forward(self, x, cos, sin):
        B, T, C = x.size() # Batch, Seq, Embed Dim
        
        # 1. Project
        q = self.q_proj(x)
        k = self.k_proj(x)
        v = self.v_proj(x)

        # 2. Reshape for multi-head attention
        # (B, T, nh, hs) -> (B, nh, T, hs)
        q = q.view(B, T, self.heads, self.head_dim).transpose(1, 2)
        k = k.view(B, T, self.heads, self.head_dim).transpose(1, 2)
        v = v.view(B, T, self.heads, self.head_dim).transpose(1, 2)

        # 3. Apply RoPE to Q and K (Rotary is applied per head)
        q, k = apply_rotary_pos_emb(q, k, cos, sin)

        # 4. Attention
        # is_causal=False because this is a bidirectional encoder
        y = F.scaled_dot_product_attention(q, k, v, is_causal=False)
        
        # 5. Reassemble
        y = y.transpose(1, 2).contiguous().view(B, T, C)
        y = self.c_proj(y)
        
        return y

**MLP**

In [9]:
class MLP(nn.Module):
    def __init__(self, config):
        super().__init__()
        self.c_fc = nn.Linear(config.embed_dim, int(config.mlp_ratio * config.embed_dim))
        self.gelu = nn.GELU(approximate='tanh')
        self.c_proj = nn.Linear(int(config.mlp_ratio * config.embed_dim), config.embed_dim)

    def forward(self, x):
        x = self.c_fc(x)
        x = self.gelu(x)
        x = self.c_proj(x)
        return x

**Hidden Transformer Block**

In [10]:
class PathwayBlock(nn.Module):
    def __init__(self, config):
        super().__init__()
        self.ln_1 = nn.LayerNorm(config.embed_dim)
        self.attn = BioMultiHeadAttention(config)
        self.ln_2 = nn.LayerNorm(config.embed_dim)
        self.mlp = MLP(config)

    def forward(self, x, cos, sin):
        # 1. Attention with RoPE
        x_norm = self.ln_1(x)
        # Pass cos/sin into attn to be applied to Q/K
        attn_out = self.attn(x_norm, cos, sin)
        x = x + attn_out

        # 2. MLP
        x = x + self.mlp(self.ln_2(x))
        return x

**Main Pathway Model**

In [11]:
@dataclass
class PathwayEncoderConfig:
    num_pathways: int = 1024 
    n_layer: int = 24 
    heads: int = 12
    embed_dim: int = 768
    mlp_ratio: float = 4.0 # Changed to float for precision

class PathwayEncoder(nn.Module):
    def __init__(self, config):
        super().__init__()
        self.config = config

        self.transformer = nn.ModuleDict(dict(
            # input projects to embedding
            input_proj = nn.Linear(1, config.embed_dim),
            
            # RoPE needs the HEAD dimension, not the full embedding dimension
            rope = RotaryEmbedding(config.embed_dim // config.heads),
            
            # transformer block
            blocks = nn.ModuleList([PathwayBlock(config) for _ in range(config.n_layer)]),
            
            # final layer norm
            ln_f = nn.LayerNorm(config.embed_dim) 
        ))

        self.apply(self._init_weights)

    def _init_weights(self, module):
        std = 0.02
        if isinstance(module, nn.Linear):
            torch.nn.init.normal_(module.weight, mean=0.0, std=std)
            if module.bias is not None:
                torch.nn.init.zeros_(module.bias)
        elif isinstance(module, nn.Embedding):
            torch.nn.init.normal_(module.weight, mean=0.0, std=std)
        
    def forward(self, x):
        # x: [Batch, Num_Pathways]
        x = x.unsqueeze(-1) # [B, N, 1]
        x = self.transformer.input_proj(x) # [B, N, Dim]
        
        # Generate RoPE cache
        # cos, sin are [Seq, Head_Dim]
        cos, sin = self.transformer.rope(x)
        cos, sin = cos.to(x.device), sin.to(x.device)
        
        for block in self.transformer.blocks:
            x = block(x, cos, sin)
            
        x = self.transformer.ln_f(x)

        return x

### Create The Student And Teacher Models

In [12]:
n_embd = 4
n_heads = 2
n_pathways = 1024
DEVICE = 'cuda' if torch.cuda.is_available() else 'cpu'

In [13]:
student = PathwayEncoder(PathwayEncoderConfig(n_layer= 1, heads=n_heads, embed_dim= n_embd))
student

PathwayEncoder(
  (transformer): ModuleDict(
    (input_proj): Linear(in_features=1, out_features=4, bias=True)
    (rope): RotaryEmbedding()
    (blocks): ModuleList(
      (0): PathwayBlock(
        (ln_1): LayerNorm((4,), eps=1e-05, elementwise_affine=True)
        (attn): BioMultiHeadAttention(
          (q_proj): Linear(in_features=4, out_features=4, bias=True)
          (k_proj): Linear(in_features=4, out_features=4, bias=True)
          (v_proj): Linear(in_features=4, out_features=4, bias=True)
          (c_proj): Linear(in_features=4, out_features=4, bias=True)
        )
        (ln_2): LayerNorm((4,), eps=1e-05, elementwise_affine=True)
        (mlp): MLP(
          (c_fc): Linear(in_features=4, out_features=16, bias=True)
          (gelu): GELU(approximate='tanh')
          (c_proj): Linear(in_features=16, out_features=4, bias=True)
        )
      )
    )
    (ln_f): LayerNorm((4,), eps=1e-05, elementwise_affine=True)
  )
)

**Teacher** Now let's make a copy for the teacher. Since the teacher does not require updates, we can update the model to not require gradient updates. 

In [14]:
teacher = copy.deepcopy(student)
teacher

PathwayEncoder(
  (transformer): ModuleDict(
    (input_proj): Linear(in_features=1, out_features=4, bias=True)
    (rope): RotaryEmbedding()
    (blocks): ModuleList(
      (0): PathwayBlock(
        (ln_1): LayerNorm((4,), eps=1e-05, elementwise_affine=True)
        (attn): BioMultiHeadAttention(
          (q_proj): Linear(in_features=4, out_features=4, bias=True)
          (k_proj): Linear(in_features=4, out_features=4, bias=True)
          (v_proj): Linear(in_features=4, out_features=4, bias=True)
          (c_proj): Linear(in_features=4, out_features=4, bias=True)
        )
        (ln_2): LayerNorm((4,), eps=1e-05, elementwise_affine=True)
        (mlp): MLP(
          (c_fc): Linear(in_features=4, out_features=16, bias=True)
          (gelu): GELU(approximate='tanh')
          (c_proj): Linear(in_features=16, out_features=4, bias=True)
        )
      )
    )
    (ln_f): LayerNorm((4,), eps=1e-05, elementwise_affine=True)
  )
)

In [15]:
for p in teacher.parameters():
    p.requires_grad = False

### Forward Pass: Teacher
Teacher encodes the Target (Case). We'll make sure to run it without gradient accumulation.  This teacher creates an embedding of the case cell

In [16]:
with torch.no_grad():
    target_latents = teacher(x_case)
target_latents

tensor([[[ 0.6921,  0.8291, -1.1328, -0.3884],
         [ 0.6299,  0.8201, -1.0173, -0.4327],
         [ 0.6148,  0.8170, -0.9911, -0.4407],
         ...,
         [ 0.4390,  0.7643, -0.7194, -0.4839],
         [ 0.3195,  0.7161, -0.5563, -0.4792],
         [ 0.6313,  0.8203, -1.0198, -0.4319]],

        [[ 0.7259,  0.8409, -1.2146, -0.3522],
         [ 0.6116,  0.8309, -0.9961, -0.4463],
         [ 0.6282,  0.8338, -1.0251, -0.4369],
         ...,
         [ 0.4953,  0.8016, -0.8107, -0.4862],
         [ 0.2510,  0.7074, -0.4759, -0.4824],
         [ 0.6102,  0.8306, -0.9937, -0.4471]]])

### Forward Pass: Student
Student encodes the Context (Control), With the student we need to accumulate gradients since we need to be able to train the model. 
> TO DO: Add masking here if you want extra difficulty, but the task Control->Treated is enough for now.

In [17]:
context_latents = student(x_control)
context_latents

tensor([[[ 0.7461,  0.8374, -1.2575, -0.3260],
         [ 0.6927,  0.8388, -1.1434, -0.3881],
         [ 0.6394,  0.8331, -1.0430, -0.4295],
         ...,
         [ 0.4558,  0.7855, -0.7509, -0.4904],
         [ 0.3595,  0.7499, -0.6168, -0.4926],
         [ 0.6091,  0.8278, -0.9900, -0.4469]],

        [[ 0.6436,  0.8253, -1.0437, -0.4252],
         [ 0.6403,  0.8247, -1.0379, -0.4272],
         [ 0.6053,  0.8178, -0.9769, -0.4462],
         ...,
         [ 0.4603,  0.7756, -0.7516, -0.4843],
         [ 0.2731,  0.6997, -0.4975, -0.4752],
         [ 0.6216,  0.8212, -1.0049, -0.4380]]],
       grad_fn=<NativeLayerNormBackward0>)

## Predictor Setup
We need to create a predictor that tries to guess Target (from the teacher) given Context + Action


**Adaptive Layer Normalization AdaLN**

In [18]:
class AdaLN(nn.Module):
    '''
    Adaptive Layer Norm for conditioning the predictor on action embeddings.
    The action vector regresses the Scale (gamma) and Shift (beta) of the normalization.
    '''
    def __init__(self, embed_dim, action_embed_dim):
        super().__init__()
        self.norm = nn.LayerNorm(embed_dim, elementwise_affine=False)
        self.action_mlp = nn.Sequential(
            nn.SiLU(),
            nn.Linear(action_embed_dim, 2 * embed_dim)
        )
        # Initialize to identity (gamma=0, beta=0 originally, effectively gamma=1 after logic)
        # Zero-init the last layer so the action starts as a "no-op" (identity)
        nn.init.zeros_(self.action_mlp[1].weight)
        nn.init.zeros_(self.action_mlp[1].bias)

    def forward(self, x, action_emb):
        # x: [Batch, Seq, Dim]
        # action_emb: [Batch, action_embed_dim]
        
        # Project action to style: [B, 2*D] -> [B, 1, 2*D]
        style = self.action_mlp(action_emb).unsqueeze(1) 
        gamma, beta = style.chunk(2, dim=-1)
        
        # Apply affine transformation based on action
        return self.norm(x) * (1 + gamma) + beta

**Predictor Block**

In [19]:
class PredictorBlock(nn.Module):
    def __init__(self, config):
        super().__init__()
        # 1. Conditioning (AdaLN) replaces standard LayerNorm
        self.ada_ln1 = AdaLN(config.embed_dim, config.action_embed_dim)
        
        # 2. Attention (Using the shared BioMultiHeadAttention)
        self.attn = BioMultiHeadAttention(config)
        
        # 3. Conditioning (AdaLN) for the MLP block
        self.ada_ln2 = AdaLN(config.embed_dim, config.action_embed_dim)
        
        # 4. MLP (Using the shared MLP)
        self.mlp = MLP(config)

    def forward(self, x, action_emb, cos, sin):
        # 1. AdaLN -> Attention (with internal RoPE) -> Residual
        x_norm = self.ada_ln1(x, action_emb)
        
        # Note: BioMultiHeadAttention handles q/k/v projection and apply_rotary_pos_emb internally
        attn_out = self.attn(x_norm, cos, sin)
        x = x + attn_out
        
        # 2. AdaLN -> MLP -> Residual
        x_norm = self.ada_ln2(x, action_emb)
        x = x + self.mlp(x_norm)
        
        return x

**Main Predictor Model**

In [20]:
@dataclass
class ACPredictorConfig:
    num_pathways: int = 1024
    n_layer: int = 6 
    heads: int = 4
    embed_dim: int = 384
    action_embed_dim: int=256 
    mlp_ratio: float = 4.0
    max_perturb: int= 2058 ## eventually try to get to a 2**N power

class ACPredictor(nn.Module):
    def __init__(self, config):
        super().__init__()
        self.config = config
        
        # Action Embedding (Discrete ID -> Vector)
        self.action_embed = nn.Embedding(config.max_perturb, config.action_embed_dim)
        
        # Learnable Queries ("Mask Tokens") for the future state
        # One query vector per pathway position
        self.mask_queries = nn.Parameter(torch.randn(1, config.num_pathways, config.embed_dim) * 0.02)
        
        # RoPE: initialized with HEAD dimension (dim // heads)
        head_dim = config.embed_dim // config.heads
        self.rope = RotaryEmbedding(head_dim)
        
        self.blocks = nn.ModuleList([
            PredictorBlock(config) for _ in range(config.n_layer)
        ])
        
        self.final_norm = AdaLN(config.embed_dim, config.action_embed_dim)
        
        self.apply(self._init_weights)

    def _init_weights(self, module):
        std = 0.02
        if isinstance(module, nn.Linear):
            torch.nn.init.normal_(module.weight, mean=0.0, std=std)
            if module.bias is not None:
                torch.nn.init.zeros_(module.bias)
        elif isinstance(module, nn.Embedding):
            torch.nn.init.normal_(module.weight, mean=0.0, std=std)

    def forward(self, context_latents, action_ids):
        """
        context_latents: [Batch, N, Dim] (From Student Encoder)
        action_ids: [Batch] (Ints)
        """
        B, N, D = context_latents.shape
        
        # 1. Embed Action
        action_emb = self.action_embed(action_ids) # [B, action_embed_dim]
        
        # 2. Construct Input: [Context, Mask_Queries]
        # We concatenate the learned queries to the context. 
        # The predictor will attend to the context to update the queries.
        queries = self.mask_queries.repeat(B, 1, 1) # [B, N, D]
        sequence = torch.cat([context_latents, queries], dim=1) # [B, 2N, D]
        
        # 3. Generate RoPE for the full sequence (2N)
        # cos, sin are [2N, Head_Dim]
        cos, sin = self.rope(sequence)
        cos, sin = cos.to(sequence.device), sin.to(sequence.device)
        
        # 4. Pass through AdaLN Blocks
        for block in self.blocks:
            sequence = block(sequence, action_emb, cos, sin)
            
        sequence = self.final_norm(sequence, action_emb)
        
        # 5. Return only the predicted part (The Queries corresponding to N..2N)
        predictions = sequence[:, N:, :] 
        return predictions

### Create the Predictor Model

In [21]:
predictor = ACPredictor(ACPredictorConfig(n_layer=1, heads=n_heads, embed_dim=n_embd, action_embed_dim=n_embd))
predictor

ACPredictor(
  (action_embed): Embedding(2058, 4)
  (rope): RotaryEmbedding()
  (blocks): ModuleList(
    (0): PredictorBlock(
      (ada_ln1): AdaLN(
        (norm): LayerNorm((4,), eps=1e-05, elementwise_affine=False)
        (action_mlp): Sequential(
          (0): SiLU()
          (1): Linear(in_features=4, out_features=8, bias=True)
        )
      )
      (attn): BioMultiHeadAttention(
        (q_proj): Linear(in_features=4, out_features=4, bias=True)
        (k_proj): Linear(in_features=4, out_features=4, bias=True)
        (v_proj): Linear(in_features=4, out_features=4, bias=True)
        (c_proj): Linear(in_features=4, out_features=4, bias=True)
      )
      (ada_ln2): AdaLN(
        (norm): LayerNorm((4,), eps=1e-05, elementwise_affine=False)
        (action_mlp): Sequential(
          (0): SiLU()
          (1): Linear(in_features=4, out_features=8, bias=True)
        )
      )
      (mlp): MLP(
        (c_fc): Linear(in_features=4, out_features=16, bias=True)
        (gelu)

### Forward Pass: Action Predictor
The Action predictor tries to guess Target encoding given Context + Action.  The context is the ouput of the Student and the action is from our tokenized data.  The goal here is tha the output attempts to match the teacher. 

In [22]:
action_id

tensor([1383,  429])

In [23]:
predicted_latents = predictor(context_latents, action_id)
predicted_latents

tensor([[[-1.3142, -0.4007,  0.3991,  1.3173],
         [ 1.5854, -0.7165, -0.8066, -0.0618],
         [-0.3736,  1.6874, -0.8535, -0.4591],
         ...,
         [ 0.4308,  1.2783, -1.1609, -0.5472],
         [-1.5897, -0.0919,  0.7758,  0.9074],
         [ 1.3575, -0.6960, -1.1524,  0.4916]],

        [[-1.3170, -0.4009,  0.3998,  1.3153],
         [ 1.5854, -0.7162, -0.8053, -0.0648],
         [-0.3752,  1.6871, -0.8519, -0.4614],
         ...,
         [ 0.4298,  1.2782, -1.1589, -0.5500],
         [-1.5928, -0.0922,  0.7768,  0.9053],
         [ 1.3573, -0.6962, -1.1510,  0.4890]]], grad_fn=<SliceBackward0>)

# Loss Calculation

We'll use L1 Loss (Mean Absolute Error) which measures the average absolute difference between predicted and actual values, making it robust to outliers.  We'll call this latent loss since we'll  use an L1 loss function within the latent space of a model to ensure the compressed representation (latent) accurately reflects the original data. 

In [24]:
loss = F.l1_loss(predicted_latents, target_latents)
loss

tensor(1.0029, grad_fn=<MeanBackward0>)

# Back Propogation

In [25]:
loss.backward()

## Gradient Clipping

In [26]:
torch.nn.utils.clip_grad_norm_(predictor.parameters(), 1.0)
torch.nn.utils.clip_grad_norm_(student.parameters(), 1.0)

tensor(6.2358)

## Update the Teacher

Recall that our teacher was created from a copy of our student. This means the weights are initiated randomly. We also didn't include any backprop for our teacher as since our loss is calculated using the action predictor and the student model.  To update the teacher we'll use a mechanism called  **Exponential Moving Average (EMA)**

In simple terms, this mechanism forces the Teacher to evolve very slowly, becoming a stable, wiser version of the Student.

Here is the breakdown of the mechanics. Mathematically, it calculates:

$Teacher_{new} = (m * Teacher_{old})+((1−m)*Student_{current})$

With a typical momentum ($m$) of 0.996, the code tells the Teacher: *"Keep 99.6% of your current weights, and mix in just 0.4% of what the Student learned in this last step."*

The reason we do this is Stability. The Student network is updating rapidly and noisily using Gradient Descent ($Param←Param−LearningRate×Gradient$). If we simply copied the Student to the Teacher (`Teacher = Student`), the target we are trying to predict would move erratically every single step, making it impossible to learn (like trying to hit a bullseye that is vibrating wildly).

By updating the Teacher this way, it effectively represents the **average** of the Student's weights over the last several thousand steps. This smooths out the noise and provides a steady, high-quality target for the Student to predict, preventing "Model Collapse" (where the model cheats by outputting zeros for everything).

TO:DO  we could also replace the teacher with a pretrained model and use this as an update mechanism. 

In [27]:
m=0.996
for param_s, param_t in zip(student.parameters(), teacher.parameters()):
    param_t.data.mul_(m).add_((1 - m) * param_s.data)

## Inference
Now we've done a walkthrough of the forward pass of the model let's see how we can run an inference test since this is an encoder model. We'll reload our tokenized data and perturbation map so that we can convert our pertubation to text and sample from our dataset. 

In [28]:
metadata_path, train_dir

(PosixPath('/Users/djemec/data/jepa/perturbation_map.json'),
 PosixPath('/Users/djemec/data/jepa/tokenized/train'))

In [29]:
# 1. Load Map (ID -> Name)
with open(metadata_path, "r") as f:
    pert_map = json.load(f)
# Invert map to: ID -> Name
id_to_name = {v: k for k, v in pert_map.items()}

**Sample Data**
We'll build a modification of our data loader which pulls a random file and a random perturbed cell example from it.  

In [30]:
def get_random_test_pair(train_dir):
    '''Grab a single real pair from a random shard'''
    files = sorted(train_dir.glob('*.npz'))
    file_path = files[np.random.randint(len(files))]
    
    with np.load(file_path) as data:
        idx = np.random.randint(data['action_ids'].shape[0])
        
        # Extract raw uint32
        control_raw = data['control'][idx]
        case_raw = data['case'][idx]
        act_id = data['action_ids'][idx]
        
    # Dequantize
    x_control = torch.tensor(control_raw.astype(np.float32)).unsqueeze(0).to(DEVICE) # [1, 1024]
    x_case = torch.tensor(case_raw.astype(np.float32)).unsqueeze(0).to(DEVICE)
    action_id = torch.tensor([act_id], dtype=torch.long).to(DEVICE)
    
    return x_control, x_case, action_id

In [31]:
x_control, x_case, action_id = get_random_test_pair(train_dir)
pert_name = id_to_name[action_id.item()]
pert_name

'YARS2'

**Get Baseline**

To start, we'll take our `teacher` model to create our baseline expected embedding for both our control and case examples.  These create our "expected" values 

In [32]:
# 2. Get Baselines (Teacher View)
# We need the Teacher to tell us where the "Control" and "Real Treated" 
# sit in the abstract latent space.
with torch.no_grad():
    z_control = teacher(x_control)       # Where the cell started
    z_case = teacher(x_case)     # Where the cell actually went

**Get Predicted**

Now we need to run our predicted model. This requires first running our control (baseline cell) through our `student`, then use the output with our `predictor` along with the perturbation to see how good our predictor is. 

In [33]:
# 3. Run The Physics Engine (Predictor)
# Student encodes context -> Predictor adds Action -> Output
with torch.no_grad():
    z_context = student(x_control)
    z_predicted = predictor(z_context, action_id) # Where the model thinks it went

### Metrics

Now we'll calculate 3 metrics:
1. **Baseline Drift** - First we have to understand how much our teacher expects the cell state to change.  If the teacher is not expecting a change then we'd want to know that and the value would be `0`.  
2. **Prediction Error** - Second we'll now check to see how far our predictor is from our expected case based on the teacher.  This is the value we'd like to minimize as best as possible. 
3. **Simulation Magnitude** - Third we'll see how much the predictor is different from the teacher based control. This value should ideally be as close to the drift value as possible since the predictor's output should share the latent space with the teacher

With these values, we can understand a few things: If the error is larger than the drift, we can infer that the model is still worse than just predicting no change. Also, by comparing drift and simulated magnitude move we can understand how close directionally our predictor is to the teacher's impact. If that is close we can see that we're improving our model.  

In [34]:
# Metric 1: Baseline Drift (How much did the drug actually change the cell?)
# If this is 0, the drug did nothing, so prediction is trivial.
drift = F.l1_loss(z_control, z_case).item()
drift

0.038677558302879333

In [35]:
# Metric 2: Prediction Error (How close is our guess to the real result?)
error = F.l1_loss(z_predicted, z_case).item()
error

1.012245535850525

In [36]:
# Metric 3: Simulation Magnitude (How much did our model decide to move the cell?)
sim_move = F.l1_loss(z_control,z_predicted).item()
sim_move

1.0195283889770508

## Conclusion
As you can see, untrained we have quite a lot of error.  Also, we're juggling 3 models and our predictor relies on a chain of two models meaning that overall we'll need a lot of training to start learning the cell physics. 

Also, looking at how this works we can see that we have a limitation on our training data currently as we're collapsing full networks down to a single number. 